### ISSUE TITLE
Using octet disc. in union results in opendds_idl compile error

### ISSUE THREAD

**Issue by jwillemsen** [objectcomputing/OpenDDS#4794]
When using `@optional` with a union opendds_idl gives an error

Test file:
```
union TestUnion_Octet switch(octet) {
  case 1:
    @optional short FirstCase;
  default:
    @optional long myinft;
};
````

Compiled using:

```
opendds_idl -Sa -St --no-dcps-data-type-warnings -Lc++11 --no-default-nested -Gxtypes-complete optional.idl
```

Results in the error

```
Error - /home/johnny/ACE/OpenDDS/bin/opendds_idl: "optional.idl", line 7: Unhandled ExprType value in generateDefaultValue
Fatal Error - Aborting
```

**Comment by jwillemsen** [comment]
DDS XTypes 1.3 says in 7.3.1.2.2 that `@optional` may be applied to `Structure Members`, but IDL4.2 doesn't mention this restriction, it only says `This annotation may be used to set optionality on any element that makes sense to be optional.`. Reported an issue to the OMG to get this better specified in IDL4, see https://issues.omg.org/issues/IDL43-100, is it or is it not legal to use `@optional` on a union member

**Comment by mitza-oci** [comment]
Based on https://issues.omg.org/issues/spec/IDL#issue-44207 the IDL spec is only concerned with general syntax of annotations and not semantics (including where they may be applied).

**Comment by jwillemsen** [comment]
So with CORBA we can say we support `@optional` for unions, but DDS can say we don't :-(

**Comment by mitza-oci** [comment]
It can probably be added to XTypes without much effort, UnionMemberFlag can have IS_OPTIONAL.

Or just "wrap" the type in a struct?

**Comment by jwillemsen** [comment]
I have reported an issue on XTypes, also exception should be added, as IDL exceptions can now be used through DDS RPC

**Comment by mitza-oci** [comment]
> So with CORBA we can say we support `@optional` for unions

Not yet, since there's no CDR / TypeObject for optional.

**Comment by jwillemsen** [comment]
> > So with CORBA we can say we support `@optional` for unions
> 
> Not yet, since there's no CDR / Typecode for optional.

Yes, not in the spec yet. I created a proof of concept for CDR in TAOX11, marshal it as a union with a boolean discriminator, that seems to work perfect, and even makes it possible to communicate with a non IDL4 CORBA implementation. Typecode/DynamicAny/InterfaceRepo proof of concept is on the todo list.

**Comment by iguessthislldo** [comment]
Support for using octet disc. was added in ACE master in https://github.com/DOCGroup/ACE_TAO/pull/2244, but the support for this in OpenDDS I started on hasn't been merged yet.

**Comment by jwillemsen** [comment]
Ok, so this are probaby two issues, octet and optional
